### PR TITLE
A more robust createStore inspired by @foxnewsnetwork

### PR DESCRIPTION
**Pull request by toranb**
As part of the feedback for PR #120 @foxnewsnetwork passed over a createStore that is more extensible (while not breaking the app itself). I took inspiration from that and my previous PR to build this one #feedbackWelcome

### CODE REVIEW

**Comment by toranb** [comment]
screen shot 2017-04-29 at 9 51 14 pm

Proof this implementation not only has a green build but won't bust your app 👍

File: app/reducers/optional.js [hunk 1]
```diff
@@ -1,7 +0,0 @@
-//this is a simple pass thru that offers an extension point
```
Comment (toranb) [inline app/reducers/optional.js]: just for reference -this was removed from ember-redux just before the 2.0 release

https://github.com/ember-redux/ember-redux/commit/be2f989cea73a802a0afaacb011dffa59cef4e09

**Comment by toranb** [comment]
I added a new commit to move the `redux-thunk` imports to app/ instead of addon/

By doing this the consuming app can truly/safely remove `redux-thunk` as a dependency and add something like `redux-saga` :)

**Comment by toranb** [comment]
Because this was a move the service diff isn't a great read - if you want to see a simple diff I pushed a patch up to [dropbox](https://dl.dropboxusercontent.com/u/716525/downloads/opensource/theDiff.patch) (worth a look if you want a simple idea of what has changed in the service)

**Comment by toranb** [comment]
I think the last remaining step here could/would? be to make `makeStoreInstance` something we import explicitly so "power users" can truly just override that part producing a store configured any way they want (like you would by hand using `react-redux` for example).

Anyone have thoughts on this? where would we export it from exactly?

**Comment by foxnewsnetwork** [comment]
`makeStoreInstance` is already merged in with the other stuff as a part of extending `Ember.Service`

Therefore, power-users can already do something like:

```javascript
import ReduxService from 'ember-redux/services/redux';

export default ReduxService.extend({
  makeStoreInstance({ middlewareConfig, enhaners, reducers }) { 
    doWhatever()
  })
```

**Comment by foxnewsnetwork** [comment]
LGTM. Also, please consider also editing over some of the howto extend from `README.markdown` stuff on my branch: https://github.com/foxnewsnetwork/ember-redux/tree/feature/addonify-redux-service#customizing-the-redux-service

File: addon/middleware/index.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+//only here as a vanilla pass thru so we can
```
Comment (brettburley) [inline addon/middleware/index.js]: What is this middleware for? Is it used at all?

File: addon/services/redux.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import Ember from 'ember';
+import redux from 'redux';
+import reducers from '../reducers/index';
```
Comment (brettburley) [inline addon/services/redux.js]: Since these imports are within the addon itself, should we actually exclude them here, and just rely on them being passed into the extended version of the service? I don't think they add any value here since we want to include the `reducers`, `enhancers`, and `middlewareConfig` from the host app anyway.

File: app/reducers/index.js [hunk 4]
```diff
@@ -0,0 +1 @@
+export {default} from 'ember-redux/reducers/index';
```
Comment (brettburley) [inline app/reducers/index.js]: Previously, did someone using `ember-redux` always have to create a `reducers/index.js` file in order to prevent their app from breaking?

File: addon/services/redux.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  );
+  return isArray(mc) ? { middleware: mc } : mc;
+}
+
+const { createStore, applyMiddleware, combineReducers, compose } = redux;
+
+const makeStoreInstance = ({middlewareConfig, reducers, enhancers}) => {
+  const { middleware, setup = () => {} } = extractMiddlewareConfig(middlewareConfig);
+  const createStoreWithMiddleware = compose(applyMiddleware(...middleware), enhancers)(createStore);
+  const reducer = typeof reducers === 'function' ? reducers : combineReducers(reducers);
```
Comment (brettburley) [inline addon/services/redux.js]: Is this intentionally including the change from #119?

File: app/reducers/index.js (hunk 4 above)
Comment (toranb) [inline app/reducers/index.js]: yes - it would throw an error letting you know the reducers/index.js was missing (not great but it also was a solid entry point into "starting using redux ..."). This improves it as we don't break out of the box (when this PR lands)

File: addon/services/redux.js (hunk 5 above)
Comment (toranb) [inline addon/services/redux.js]: only because I'm 99% sure I'll merge that PR in the next 24 hours and I'll rebase this one right after (sorry for the confusion here - mostly trying to work conflict free)

File: addon/services/redux.js (hunk 3 above)
Comment (toranb) [inline addon/services/redux.js]: @foxnewsnetwork ^^ any reason we can't/or shouldn't do this? on the surface it seems redundant as @brettburley mentioned

File: addon/middleware/index.js (hunk 2 above)
Comment (toranb) [inline addon/middleware/index.js]: because the addon/services/redux file calls out an explicit "import" for middleware I need this file present. If I include the `redux-thunk` version here (instead of app) the end user is then required to have redux-thunk installed even if they opt-out and use redux- saga instead

Comment (brettburley) [inline addon/middleware/index.js]: Actually this file is an artifact of https://github.com/ember-redux/ember-redux/pull/121#discussion_r114086961, right?

Also, depending on the resolution of that comment, could this file be simply an `export default [];` or does that not work/am I missing something else?

**Comment by toranb** [comment]
@brettburley I've rebased this and cut the addon/middleware down to just an empty array as you suggested. I still need to pull in some additional documentation for this as @foxnewsnetwork suggested above ^^

File: addon/reducers/index.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+// const foobar = (state, action) => {
+//   //reducer code here
+// };
+//
+// export default combineReducers({
+//   foobar
+// })
+
+export default function(combine) {
+  return (state, action) => {
```
Comment (toranb) [inline addon/reducers/index.js]: @brettburley should I add some type of console log or something here to alert the consuming app that they need to write a custom reducer?

File: addon/services/redux.js (hunk 5 above)
Comment (toranb) [inline addon/services/redux.js]: @brettburley I've rebased this and it should be down to a single commit now :)

File: addon/reducers/index.js (hunk 6 above)
Comment (brettburley) [inline addon/reducers/index.js]: That seems like a reasonable approach. I've been using `Ember.warn` for those purposes (that way, it's stripped from production builds).

Also, doesn't this reducer completely fail as-is, since there's no combine being passed in? I think this file needs to just export a reducer, like:
```
export default function (state, action) {
 return state;
};
```

**Comment by toranb** [comment]
Okay I've got this rebased - still needs some docs love (copy/paste of what Thomas linked us to above). I need to see how this looks/feels in a real app so that's on me this coming week

**Comment by bstro** [comment]
Any estimate on when this sweet PR might land? Anything I can do to help?

**Comment by toranb** [comment]
@bstro The only real blocker at the moment is my internal debate over the "breaking change" this introduced. The problem is that previously we wouldn't include a reducer for you (meaning an error message would pop up when you ran the build w/out `reducers/index.js`

We are planning to cut v3.0 in early July and this is likely the 2nd PR that will land officially (in that case any of the fears about a breaking api change will have a major version number to land under). 

I don't have anything else planned until that release next month but we could use some examples/ docs/ for people who want to extend this like yourself. Would you want to write up a small section for the docs that covers this in more detail?

**Comment by bstro** [comment]
Ah, so this won't land in v2.x? In order to have these changes, I'd also need to upgrade our version of node+ember+ember-cli?

```
ember-cli: 2.4.3
node: 4.8.2
os: darwin x64
```

**Comment by toranb** [comment]
@bstro It's possible we could land this in 2x -worth a debate on the "breaking change" I mentioned above. That said, in the discussion about 3X I showed a graph after node 8 was released recently and asked about what may be blocking you from taking on node 6 as a build time dependency.

You can still use ember-cli v2.4.3 w/ node 6.10 and that will be supported in both 2X and 3X of `ember-redux`. The problem arises when you need to build your ember-cli v2.4.3 app with node 4 (you will find a break as a result of the upgrade from babel 5 to babel 6).

What is blocking you/ your team from adopting node 6 as a build time dependency today?

**Comment by brettburley** [comment]
@toranb I don't think it's a _breaking_ change since there aren't any meaningful apps with a missing `reducers/index.js` file. I don't think this PR changes the API in any way for existing apps, and only changes behaviors for new apps, so wouldn't necessitate a major version change because there is no work an app owner would have to do. I also think adding the runtime warning is probably sufficient warning for the new app case.

I'd strongly support landing this in the current major version (with the caveats of my previous feedback). :+1:

File: addon/services/redux.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  const store = createStoreWithMiddleware(reducer);
+  setup(store);
+  return store;
+};
+
+export default Ember.Service.extend({
+  middlewareConfig,
```
Comment (brettburley) [inline addon/services/redux.js]: Is there a reason this field is named `middlewareConfig` rather than just `middleware` (matching the pattern of `reducers`, `enhancers`)? Just want to be careful with the naming since these will become part of the public API when overriding the default implementation here.

**Comment by bstro** [comment]
@toranb I'm going to spike this week on upgrading our Node to 6 LTS. Some promising progress so far.

**Comment by toranb** [comment]
@brettburley it sounds like we agree that this change does not introduce a real/ functional breaking change of any kind. 

@brettburley I've updated the name you called out `middlewareConfig` to instead match the others (plural) `middlewares`. One question with that name change - is it correct? Meaning, we could bolt on 2+ middleware (one for logging/ one for async work/ etc)

The last step before we merge this into the 2x series is for me to find a way to test this from the dummy app (like we did w/ patchMiddleware). If that's not possible/or easily possible I'd like to test this manually with ember-cli v2.4.3 and v2.13.2

File: addon/services/redux.js (hunk 3 above)
Comment (foxnewsnetwork) [inline addon/services/redux.js]: IIRC importing them there was for backwards compatibility reasons so that users of older version of ember-redux wouldn't have to change up the way they're importing... but if we're bumping major versions for these sets of changes, I think @brettburley 's approach is valid.

**Comment by toranb** [comment]
@foxnewsnetwork @brettburley @bstro 

I threw together a working ember-cli v2.4.3 app using this fork and customized the redux store creation. Looking for feedback on this before we merge it next week

https://github.com/toranb/configme243/commit/647145fa679c58598a352bc6a25a9b93607270d8

**Comment by bstro** [comment]
Your 2.4.3 example looks consistent with how I've been using it! It feels pretty good to me. Nothing to critique on my end.

**Comment by toranb** [comment]
@bstro one issue I've found today is that I can destructure enhancers just fine but not the reducers for some reason. The below doesn't blow up, but the reducers don't fire as they do when I wire it up myself. I'll poke around at this until I can get a config that truly provides reducers, middlewares, and enchancers.

```js
export default ReduxService.extend({
  makeStoreInstance({reducers, enhancers}) {
    const middleware = applyMiddleware(sagaMiddleware);
    const createStoreWithMiddleware = compose(middleware, enhancers)(createStore);
    const store = createStoreWithMiddleware(reducers);
    sagaMiddleware.run(root);
    return store;
  }
});
```

**Comment by bstro** [comment]
@toranb now that you mention it, I recall running into that too. I had to manually import my reducers.

**Comment by toranb** [comment]
@bstro I believe the variables we destructure in `makeStoreInstance` reference the "default" imports that ship with `ember-redux` itself (meaning the reducer/middleware/enchancer we ship that do next to nothing).

@foxnewsnetwork is this how you'd expect this to work? Not a complete deal breaker obviously, just trying to understand how those vars coming into `makeStoreInstance` should work/ what the developer experience is like using them

**Comment by bstro** [comment]
@toranb Doesn't that make the params to makeStoreInstance effectively useless?

**Comment by bstro** [comment]
@toranb I don't think its a big deal to manually import reducers and enhancers...

**Comment by toranb** [comment]
@bstro I'm personally good with this myself (just wasn't sure if that was the expectation when I took my first stab at extending it). This does have the added benefit of "no breaking change" for people in the 2X series. Plus it still allows us to provide a default middleware using `redux-thunk` without the user even knowing it was required (often the very first question people new to redux ask after "hello world" when they attempt something async).

When I look at the source in `/app/services/redux.js` I see that we are importing all 3 at the top and pass those in like so ...

```js
import ReduxService from 'ember-redux/services/redux';
import reducers from '../reducers/index';
import enhancers from '../enhancers/index';
import middlewares from '../middleware/index';

export default ReduxService.extend({ reducers, enhancers, middlewares });
```

Making me think it's logical that anyone extending this would then be required to import reducers/middlewares/enhancers or whatever they need to create the redux store manually.

TL;DR

I think this all makes sense and with a legit documentation section (under configuration) we could easily show how one might bolt on a custom middleware, or reducer, or enhancer.

I'll give @brettburley and @foxnewsnetwork time next week to think this over and provide feedback. I'm often in my own little world ;) so I'll step back and let them reflect on this (now) massive PR (sorry in advance guys)

**Comment by toranb** [comment]
@bstro one last look at how we could show this in action. I took a step back and found this worked perfectly (as I import the reducers/enhancers myself and construct the saga middleware by hand)

```js
import redux from 'redux';
import ReduxService from 'ember-redux/services/redux';
import createSagaMiddleWare from 'redux-saga';
import reducers from '../reducers/index';
import enhancers from '../enhancers/index';
import root from '../sagas/index';

const { createStore, applyMiddleware, compose } = redux;

const sagaMiddleware = createSagaMiddleWare();

const makeStoreInstance = ({reducers, enhancers}) => {
  const middleware = applyMiddleware(sagaMiddleware);
  const createStoreWithMiddleware = compose(middleware, enhancers)(createStore);
  const store = createStoreWithMiddleware(reducers);
  sagaMiddleware.run(root);
  return store;
};

export default ReduxService.extend({
  reducers,
  enhancers,
  makeStoreInstance
});
```

**Comment by foxnewsnetwork** [comment]
As far as I can tell, this LGTM; just be sure to at least bump minor (if not major) version so that we can always patch if users come up with unforeseen problems

**Comment by bstro** [comment]
your usage example looks great to me. +1 -- thanks for your work on this! all your work has been super helpful for me and several people on my team.

File: addon/reducers/index.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// const foobar = (state, action) => {
+//   //reducer code here
+// };
+//
+// export default combineReducers({
+//   foobar
+// })
+import Ember from 'ember';
+
+export default function(combine) {
```
Comment (brettburley) [inline addon/reducers/index.js]: I'm still a bit confused with what exactly this is exporting here, so I'm sure I'm missing something. Should this not just be a no-op reducer, e.g.:
```
export default function(state) {
  Ember.warn(...);
  return state;
});
```
otherwise, lgtm

Comment (toranb) [inline addon/reducers/index.js]: Ah, I remember finding this in some redux source at one point. Because it's not actually intended for use we could default to the more simplistic bersuon you recommended.

When I'm back from vacation early next week I'll update this.

**Comment by toranb** [comment]
@bstro @brettburley @foxnewsnetwork this has been released as part of ember-redux v2.10

**Comment by bstro** [comment]
Nice!!
